Dependencies/xr/Source/ARKit/XR.mm

```diff
@@ -0,0 +1,782 @@
+#include <XR.h>
+
+#import <UIKit/UIKit.h>
+#import <ARKit/ARKit.h>
+#import <ARKit/ARConfiguration.h>
+#import <MetalKit/MetalKit.h>
+
+@interface SessionDelegate : NSObject <ARSessionDelegate>
+@property (readonly) id<MTLTexture> cameraTextureY;
+@property (readonly) id<MTLTexture> cameraTextureCbCr;
+@end
+
+namespace
+{
+    typedef struct
+    {
+        vector_float2 position;
+        vector_float2 uv;
+        vector_float2 cameraUV;
+    } XRVertex;
+
+    static XRVertex vertices[] =
+    {
+        // 2D positions, UV         camera UV
+        { { -1, -1 },   { 0, 0 },   { 0, 0} },
+        { { -1, 1 },    { 0, 1 },   { 0, 0} },
+        { { 1, -1 },    { 1, 0 },   { 0, 0} },
+        { { 1, 1 },     { 1, 1 },   { 0, 0} },
+    };
+}
+
+/**
+ Implementation of the ARSessionDelegate interface for more info see: https://developer.apple.com/documentation/arkit/arsessiondelegate
+ */
+@implementation SessionDelegate
+{
+    std::vector<xr::System::Session::Frame::View>* activeFrameViews;
+    CVMetalTextureCacheRef textureCache;
+    id<MTLTexture> _cameraTextureY;
+    id<MTLTexture> _cameraTextureCbCr;
+    CVPixelBufferRef frameBuffer;
+    
+    UIInterfaceOrientation cameraUVReferenceOrientation;
+    CGSize cameraUVReferenceSize;
+}
+
+/**
+ Initializes this session delgate with the given frame views and metal graphics context.
+ */
+- (id)init:(std::vector<xr::System::Session::Frame::View>*)activeFrameViews metalContext:(id<MTLDevice>)graphicsContext
+{
+    self = [super init];
+    self->activeFrameViews = activeFrameViews;
+    
+    CVReturn err = CVMetalTextureCacheCreate(kCFAllocatorDefault, nil, graphicsContext, nil, &textureCache);
+    if (err)
+    {
+        throw std::runtime_error{"Unable to create Texture Cache"};
+    }
+    
+    return self;
+}
+
+/**
+ Returns the orientation of the app based on the current status bar orientation.
+*/
+- (UIInterfaceOrientation)orientation
+{
+    auto sharedApplication = [UIApplication sharedApplication];
+    auto window = sharedApplication.windows.firstObject;
+    if (@available(iOS 13.0, *)) {
+        if (window && window.windowScene)
+        {
+            return window.windowScene.interfaceOrientation;
+        }
+    }
+
+    return [[UIApplication sharedApplication] statusBarOrientation];
+}
+
+/**
+ Returns the viewportSize as determined by the texture size of the first active frame view.
+*/
+- (CGSize)viewportSize
+{
+    auto frameSize = activeFrameViews->front().ColorTextureSize;
+    return CGSizeMake(frameSize.Width, frameSize.Height);
+}
+
+/**
+ Called every frame during the active ARKit session.  Updates the AR Camera texture, and Camera pose. If a size change is detected also sets the UVs, and FoV values.
+*/
+- (void)session:(ARSession *)session didUpdateFrame:(ARFrame *)frame
```

Comment:
Maybe update the comment to note that this is part of the `ARSessionDelgate` protocol?